<h2>Glossary</h2>

1. **problem solving:** The process of formulating a problem, finding a solution, and expressing the solution.
2. **high-level language:** A programming language like Python that is designed to be easy for humans to read and write.
3. **low-level language:** A programming language that is designed to be easy for a computer to execute; also called “machine language” or “assembly language.”
4. **portability:** A property of a program that can run on more than one kind of computer.
5. **interpret:** To execute a program in a high-level language by translating it one line at a time.
6. **compile:** To translate a program written in a high-level language into a low-level language all at once, in preparation for later execution.
7. **source code:** A program in a high-level language before being compiled.
8. **object code:** The output of the compiler after it translates the program.
9. **executable:** Another name for object code that is ready to be executed
10. **prompt:** Characters displayed by the interpreter to indicate that it is ready to take input from the user.


11. **script:** A program stored in a file (usually one that will be interpreted).
12. **interactive mode:** A way of using the Python interpreter by typing commands and expressions at the prompt.
13. **script mode:** A way of using the Python interpreter to read and execute statements in a script.
14. **program:** A set of instructions that specifies a computation.
15. **algorithm:** A general process for solving a category of problems.
16. **bug:** An error in a program.
17. **debugging:** The process of finding and removing any of the three kinds of programming errors.
18. **syntax:** The structure of a program.
19. **syntax error:** An error in a program that makes it impossible to parse (and therefore impossible to interpret).
20. **exception:** An error that is detected while the program is running.

21. **semantics:** The meaning of a program.
22. **semantic error:** An error in a program that makes it do something other than what the programmer intended.
23. **natural language:** Any one of the languages that people speak that evolved naturally.
24. **formal language:** Any one of the languages that people have designed for specific purposes, such as representing mathematical ideas or computer programs; all programming languages are formal languages.
25. **token:** One of the basic elements of the syntactic structure of a program, analogous to a word in a natural language.
26. **parse:** To examine a program and analyze the syntactic structure.
27. **print statement:** An instruction that causes the Python interpreter to display a value on the screen.

<h2>Exercises</h2>

**Exercise 1.2.** Use a web browser to go to the Python website http://python.org . This page
contains information about Python and links to Python-related pages, and it gives you the ability to
search the Python documentation.
For example, if you enter print in the search window, the first link that appears is the documentation of the print statement. At this point, not all of it will make sense to you, but it is good to know
where it is.

**Exercise 1.3.** Start the Python interpreter and type help() to start the online help utility. Or you
can type help('print') to get information about the print statement.
If this example doesn’t work, you may need to install additional Python documentation or set an
environment variable; the details depend on your operating system and version of Python.

In [1]:
help('print')

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



In [3]:
help('str.isalnum')

Help on method_descriptor in str:

str.isalnum = isalnum(self, /)
    Return True if the string is an alpha-numeric string, False otherwise.
    
    A string is alpha-numeric if all characters in the string are alpha-numeric and
    there is at least one character in the string.



**Exercise 1.4.** Start the Python interpreter and use it as a calculator. Python’s syntax for math
operations is almost the same as standard mathematical notation. For example, the symbols +, - and
/ denote addition, subtraction and division, as you would expect. The symbol for multiplication is
*.
If you run a 10 kilometer race in 43 minutes 30 seconds, what is your average time per mile? What
is your average speed in miles per hour? (Hint: there are 1.61 kilometers in a mile).

In [6]:
ten_km_time = 43.5  # We run 10km in 43.5 minutes

In [7]:
# To find average time per mile, first divide 10 km into mile
ten_km_to_mile = 10 / 1.61
ten_km_to_mile

6.211180124223602

In [8]:
# Therefore we run 6.2 miles in 43.5 minutes, 
# now to find how much we run in a mile we simply divide
one_mile_time = ten_km_time / ten_km_to_mile
one_mile_time

7.003500000000001

Therefore the average time we run a mile is 7 minutes

In [12]:
# To find average speed in miles per hours...
# speed is distance over time... 
# Therefore let's fisrt find howmany miles (distance) we run in an hour

time = 60  # Time is 60 minutes for an hour
distance_in_miles_per_hour = time / one_mile_time
distance_in_miles_per_hour

8.567144998929106

In [13]:
ave_speed_in_miles_per_hour = distance_in_miles_per_hour / 60
ave_speed_in_miles_per_hour

0.14278574998215177

Therefore the average speed we run in miles per hour is 0.14 miles per hour